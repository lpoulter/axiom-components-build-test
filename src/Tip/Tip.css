:root {
  /**
   * The designs of the medium tip is 1.5rem from corner to corner. To calculate
   * the pre-rotated square we use `Math.sqrt((1.5rem**2)/2)`
   */
  --cmp-tip-size--medium: 1.0606601717798212rem;

  --cmp-tip-size--small: 0.75rem;
}

.ax-tip__content {
  background-color: var(--color-theme-background);
  /* stylelint-disable */
  -webkit-mask-image: -webkit-radial-gradient(white, black);
  /* stylelint-enable */
}

.ax-tip--shadow .ax-tip__content {
  background-color: transparent;
}

.ax-tip__arrow--small {
  width: var(--cmp-tip-size--small);
  height: var(--cmp-tip-size--small);
  margin: var(--cmp-tip-size--small);

  @nest .ax-tip--top & { bottom: calc(var(--cmp-tip-size--small) * -0.5); }
  @nest .ax-tip--right & { left: calc(var(--cmp-tip-size--small) * -0.5); }
  @nest .ax-tip--bottom & { top: calc(var(--cmp-tip-size--small) * -0.5); }
  @nest .ax-tip--left & { right: calc(var(--cmp-tip-size--small) * -0.5); }
}

.ax-tip__arrow--medium {
  width: var(--cmp-tip-size--medium);
  height: var(--cmp-tip-size--medium);
  margin: var(--cmp-tip-size--medium);

  @nest .ax-tip--top & { bottom: calc(var(--cmp-tip-size--medium) * -0.5); }
  @nest .ax-tip--right & { left: calc(var(--cmp-tip-size--medium) * -0.5); }
  @nest .ax-tip--bottom & { top: calc(var(--cmp-tip-size--medium) * -0.5); }
  @nest .ax-tip--left & { right: calc(var(--cmp-tip-size--medium) * -0.5); }
}

.ax-tip__arrow {
  position: absolute;
  transform: rotate(45deg);
  background-color: var(--color-theme-background);
  pointer-events: none;

  @nest .ax-tip--faded & { background-color: var(--color-theme-text); }
  @nest .ax-tip--success & { background-color: var(--color-ui-success); }
  @nest .ax-tip--warning & { background-color: var(--color-ui-warning); }
  @nest .ax-tip--error & { background-color: var(--color-ui-error); }
  @nest .ax-tip--info & { background-color: var(--color-ui-info); }
  @nest .ax-tip--shade-1 & { background-color: var(--color-theme-background--shade-1); }
  @nest .ax-tip--shade-2 & { background-color: var(--color-theme-background--shade-2); }
  @nest .ax-tip--shade-3 & { background-color: var(--color-theme-background--shade-3); }
  @nest .ax-tip--shade-4 & { background-color: var(--color-theme-background--shade-4); }
  @nest .ax-tip--carbon & { background-color: var(--color-ui-carbon); }
  @nest .ax-tip--white & { background-color: var(--color-ui-white); }

  @nest .ax-tip--shadow & {
    box-shadow: var(--drop-shadow-theme-border), var(--drop-shadow-theme-elevation--x2);
  }
}

.ax-tip--top .ax-tip__arrow--small {
  transform: translateY(var(--cmp-tip-size--small)) rotate(45deg);
}

.ax-tip--top .ax-tip__arrow--medium {
  transform: translateY(var(--cmp-tip-size--medium)) rotate(45deg);
}

.ax-tip--right .ax-tip__arrow--small {
  transform: translateX(calc(var(--cmp-tip-size--small) * -1)) rotate(45deg);
}

.ax-tip--right .ax-tip__arrow--medium {
  transform: translateX(calc(var(--cmp-tip-size--medium) * -1)) rotate(45deg);
}

.ax-tip--bottom .ax-tip__arrow--small {
  transform: translateY(calc(var(--cmp-tip-size--small) * -1)) rotate(45deg);
}

.ax-tip--bottom .ax-tip__arrow--medium {
  transform: translateY(calc(var(--cmp-tip-size--medium) * -1)) rotate(45deg);
}

.ax-tip--left .ax-tip__arrow--small {
  transform: translateX(var(--cmp-tip-size--small)) rotate(45deg);
}

.ax-tip--left .ax-tip__arrow--medium {
  transform: translateX(var(--cmp-tip-size--medium)) rotate(45deg);
}
