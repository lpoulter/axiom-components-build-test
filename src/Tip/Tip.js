import PropTypes from 'prop-types';
import React, { Component } from 'react';
import classnames from 'classnames';
import Base from '../Base/Base';
import './Tip.css';

export default class Tip extends Component {
  static propTypes = {
    /** optional property which should be provided in case positioning of the tip is done using popperJs. If this value is true, the value of position prop will be ignored. */
    arrowRef: PropTypes.func,
    /** The content on which the tip should be placed. */
    children: PropTypes.node,
    /** Background color for the tip */
    color: PropTypes.oneOf(['success', 'warning', 'error', 'info', 'shade-1', 'shade-2', 'shade-3', 'shade-4', 'carbon', 'white']),
    /** The direction at which the Tip should be pointing. The directions are opposite, for example,
     * if the arrow should be placed on the 'left', its direction prop should be 'right'. The same applies for 'top' and 'bottom'. */
    direction: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
    /** The position of the Tip along the baseline, can be any valid css units including px/rem/percentage. Should be adjusted according ot the use case */
    position: PropTypes.string,
    /** The Tip can be configured to have a shadow effect. This value is true by default */
    shadow: PropTypes.bool,
    size: PropTypes.oneOf(['small', 'medium']),
  };

  static defaultProps = {
    shadow: true,
    direction: 'top',
    size: 'medium',
  };

  getArrowStyles = () => {
    const { position, direction } = this.props;

    switch (direction) {
    case 'top':
    case 'bottom':
      return { left: position };
    case 'left':
    case 'right':
      return { top: position };
    }
  };

  render() {
    const { arrowRef, color, direction, position, children, shadow, size, ...rest } = this.props;

    const classes = classnames(`ax-tip--${direction}`, {
      [`ax-tip--${color}`]: color,
      ['ax-tip--shadow']: shadow,
    });

    const arrowClasses = classnames('ax-tip__arrow', {
      [`ax-tip__arrow--${size}`]: size,
    });

    return (
      <Base className={ classes } { ...rest }>
        <span className={ arrowClasses } ref={ arrowRef } style={ position ? this.getArrowStyles() : {} } />
        <Base className="ax-tip__content">{ children }</Base>
      </Base>
    );
  }
}
