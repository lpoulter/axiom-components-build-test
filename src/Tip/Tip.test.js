import React from 'react';
import renderer from 'react-test-renderer';
import Tip from './Tip';
import Base from '../Base/Base';


const getComponent = (props = {}) =>
  <Tip { ...props }>
    <Base>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in
        risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices
        iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada
        fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus
        finibus, et facilisis risus porta.
    </Base>
  </Tip>;

describe('Tip', () => {
  it('renders with defaultProps', () => {
    const tree = renderer.create(getComponent()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with arrowRef', () => {
    const arrowRef = () => {};
    const tree = renderer.create(getComponent({ arrowRef : arrowRef() })).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without shadow', () => {
    const tree = renderer.create(getComponent({ shadow : false })).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders with color', () => {
    ['success', 'warning', 'error', 'info', 'shade-1', 'shade-2', 'shade-3', 'shade-4']
      .forEach((color) => {
        const component = getComponent({ color });
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
      });
  });

  it('renders with position', () => {
    const component = getComponent({ position: '1rem' });
    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });


  describe('renders with direction', () => {
    ['top', 'bottom', 'right', 'left'].forEach((direction) => {
      it(direction, () => {
        const component = getComponent({ direction });
        const tree = renderer.create(component).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });
  });
});
