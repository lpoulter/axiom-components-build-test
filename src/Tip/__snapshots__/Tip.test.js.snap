// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div
  className="ax-tip--top ax-tip--success ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[` 2`] = `
<div
  className="ax-tip--top ax-tip--warning ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[` 3`] = `
<div
  className="ax-tip--top ax-tip--error ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[` 4`] = `
<div
  className="ax-tip--top ax-tip--info ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[` 5`] = `
<div
  className="ax-tip--top ax-tip--shade-1 ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[` 6`] = `
<div
  className="ax-tip--top ax-tip--shade-2 ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[` 7`] = `
<div
  className="ax-tip--top ax-tip--shade-3 ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[` 8`] = `
<div
  className="ax-tip--top ax-tip--shade-4 ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders with arrowRef 1`] = `
<div
  className="ax-tip--top ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders with defaultProps 1`] = `
<div
  className="ax-tip--top ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders with direction bottom 1`] = `
<div
  className="ax-tip--bottom ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders with direction left 1`] = `
<div
  className="ax-tip--left ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders with direction right 1`] = `
<div
  className="ax-tip--right ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders with direction top 1`] = `
<div
  className="ax-tip--top ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders with position 1`] = `
<div
  className="ax-tip--top ax-tip--shadow"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={
      Object {
        "left": "1rem",
      }
    }
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;

exports[`Tip renders without shadow 1`] = `
<div
  className="ax-tip--top"
>
  <span
    className="ax-tip__arrow ax-tip__arrow--medium"
    style={Object {}}
  />
  <div
    className="ax-tip__content"
  >
    <div
      className=""
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vestibulum in risus sit amet dolor tempus ornare ac eget dolor. Pellentesque ultrices iaculis nunc, tempor tempus magna vehicula ut. Interdum et malesuada fames ac ante ipsum primis in faucibus. Aliquam iaculis quam ut metus finibus, et facilisis risus porta.
    </div>
  </div>
</div>
`;
